src/core/service.c

```diff
@@ -1358,10 +1372,69 @@ static int service_collect_fds(
                 rfd_names[n_fds] = NULL;
         }
 
+        if (s->open_files) {
+                size_t n_fds = rn_socket_fds + s->n_fd_store;
+
+                LIST_FOREACH(open_files, f, s->open_files) {
+                        struct stat st;
+                        _cleanup_close_ int fd = -1;
+
+                        if (stat(f->path, &st) < 0)
+                                return log_error_errno(errno, "Failed to stat %s: %m", f->path);
```

Comment:
hmm, please open the fd with O_PATH first, then use fstat() on the opened fd, determine the type, and then either use fd_reopen() to make it a proper fd, or use connect() with an AF_UNIX path of /proc/self/fd/$FD to connect to it. That removes the raciness.